inspirehep/modules/orcid/tasks.py

```diff
@@ -185,40 +188,59 @@ def attempt_push(orcid, rec_id, oauth_token):
         rec_id(int): inspire record's id to push to ORCID.
         oauth_token(string): orcid token.
     """
-    put_code = get_putcode_from_redis(orcid, rec_id)
+    put_code, hashed = get_putcode_and_hash_from_redis(orcid, rec_id)
 
     if not put_code:
-        record_put_codes = get_author_putcodes(orcid, oauth_token)
+        recache_all_author_putcodes(orcid, oauth_token)
+        put_code, hashed = get_putcode_and_hash_from_redis(orcid, rec_id)
 
-        for fetched_rec_id, fetched_put_code in record_put_codes:
-            store_record_in_redis(orcid, fetched_rec_id, fetched_put_code)
+    new_code, new_hash = push_record_with_orcid(
+        recid=str(rec_id),
+        orcid=orcid,
+        oauth_token=oauth_token,
+        put_code=put_code,
+        old_hash=hashed,
+    )
 
-        put_code = get_putcode_from_redis(orcid, rec_id)
+    if new_code != put_code:
+        store_record_in_redis(orcid, rec_id, new_code, new_hash)
 
-    new_code = push_record_with_orcid(str(rec_id), orcid, oauth_token, put_code)
 
-    if not put_code:
-        store_record_in_redis(orcid, rec_id, new_code)
+def recache_all_author_putcodes(orcid, oauth_token):
+    """Fetch all putcodes from ORCID and cache them.
+
+    Args:
+        orcid(string): an orcid identifier.
+        oauth_token(string): orcid token.
+    """
+    record_put_codes = get_author_putcodes(orcid, oauth_token)
 
+    for fetched_rec_id, fetched_put_code in record_put_codes:
+        store_record_in_redis(orcid, fetched_rec_id, fetched_put_code, None)
 
-def store_record_in_redis(orcid, rec_id, put_code):
+
+def store_record_in_redis(orcid, rec_id, put_code, hashed):
     """Store the entry <orcid:recid, value> in Redis.
 
     Args:
         orcid(string): the author's orcid.
         rec_id(int): inspire record's id pushed to ORCID.
         put_code(string): the put_code used to push the record to ORCID.
+        hash(Union[string, NoneType]): hashed ORCID record
     """
     with redis_locking_context('orcid_push') as r:
         orcid_recid_key = get_orcid_recid_key(orcid, rec_id)
-        r.set(orcid_recid_key, put_code)
+        r.set(orcid_recid_key, dumps([put_code, hashed]))
```

Comment:
Why using JSON rather than the native Redis hash data type?

F.i.:
```
127.0.0.1:6379> HMSET mykey putcode 123 hash 456
OK
127.0.0.1:6379> HGETALL mykey
1) "putcode"
2) "123"
3) "hash"
4) "456"
127.0.0.1:6379> HGET mykey putcode
"123"
127.0.0.1:6379> HGET mykey hash
"456"
```

See: https://redis.io/topics/data-types

Reply:
Done :+1: